uavcan_gui_tool/widgets/node_properties.py

```diff
@@ -280,20 +280,48 @@ def _do_firmware_update(self):
         remote_fw_file = os.path.basename(fw_file)
         logger.info('Firmware file remote path: %r', remote_fw_file)
 
+        deferred_request_handle = None
+        node_status_handle = None
+        num_remaining_requests = 4
+
+        def success():
+            nonlocal deferred_request_handle
+            nonlocal node_status_handle
+
+            if deferred_request_handle is not None:
+                deferred_request_handle.remove()
+                deferred_request_handle = None
+            if node_status_handle is not None:
+                node_status_handle.remove()
+                node_status_handle = None
+
         # Sending update requests
         def on_response(e):
+            nonlocal deferred_request_handle
+
+            assert deferred_request_handle is None
+
+            deferred_request_handle = self._node.defer(2, send_request)
+
             if e is None:
                 self.window().show_message('One of firmware update requests has timed out')
             else:
                 logger.info('Firmware update response: %s', e.response)
                 self.window().show_message('Firmware update response: %s', e.response)
-            if e is None or e.response.error != e.response.ERROR_IN_PROGRESS:
-                self._node.defer(2, send_request)
 
-        num_remaining_requests = 4
+                if e.response.error == e.response.ERROR_IN_PROGRESS:
+                    success()
+
+        def on_node_status(e):
+            if e.transfer.source_node_id == self._target_node_id and e.message.mode == e.message.MODE_SOFTWARE_UPDATE and e.message.health < e.message.HEALTH_ERROR:
```

Comment:
Lines should not be longer than 120 characters, this one is noncompliant. See https://kb.zubax.com/x/_oAh